REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd


In [105]:
import numpy as np

In [2]:
bus = pd.read_csv('businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('inspections.csv')

In [3]:
bus.head(5)

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head(5)

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [5]:
len(bus)

6315

#### Question 2b: How many unique business IDs are there?  

In [173]:
len(bus["name"].unique())


5692

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [7]:
bus["group_by_for_2c"] = 1


In [14]:
grouped_0 = bus[bus["city"].isin(["San Francisco"])]


business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  group_by_for_2c  
0     37.791116 -122.403816  +14154217044                1  
1     37.786848 -122.421547  +14157763262                1  
2     37.792888 -122.403135  +14156779494                1  
3     37.807155 -122.419004           NaN                1  
4     37.747114 -122.413641  +14156415051                1  
...         ...         ...           ...              ...  
6310        NaN         NaN           NaN                1  
6311        NaN         NaN  +14154945048                1  
6312        NaN         NaN  +14158606036                1  
6313        NaN         NaN  +14153331878                1  
6314        NaN         NaN  +14158881393                1  

[6315 rows x 10 columns]

In [19]:
grouped_0 = bus[bus["city"].isin(["San Francisco"])]

grouped = grouped_0[['group_by_for_2c', 'name', "city"]].groupby('name')
grouped_1 = grouped.sum()
grouped_1 = grouped_1.sort_values(by=['group_by_for_2c'])
grouped_1 = grouped_1.reset_index()


In [36]:
grouped_1.tail(5)

name  group_by_for_2c
5687                   WALGREENS               11
5688  SAN FRANCISCO SOUP COMPANY               11
5689                   MCDONALDS               12
5690         PEET'S COFFEE & TEA               24
5691            STARBUCKS COFFEE               72

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [53]:
dataTypeSeries = bus["postal_code"].dtypes
dataTypeSeries
##Not really sure why it says just O, but it is short for object

dtype('O')

#### Question 3b

What are the unique values of postal_code?

In [55]:
bus["postal_code"].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [56]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [57]:
vb = bus[bus['postal_code'].isin(validZip)]

In [87]:
#x = len(vb["business_id"].unique())
#type(x)

int

In [88]:
#y = len(vb)
#type(y)

int

In [89]:
#def unique_business():
#    if len(vb["business_id"].unique()) = len(vb) is True:
#        return(len(vb))
#    else:
#        print("we have some work to do")

SyntaxError: invalid syntax (<ipython-input-89-715b47adedf3>, line 2)

In [184]:
# I'm not sure why the 
# function I made above doesn't work, any help is appreciated
vb["name"].nunique()


5438

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [93]:
len(vb[pd.isnull(vb['longitude'])])

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [113]:
df_4b = vb
#np.where(condition, value if condition is true, value if condition is false)
#df['hasimage'] = np.where(df['photos']!= '[]', True, False)

df_4b["null_lon"] = np.where(df_4b["longitude"] > -200, 1, 0)
df_4b["not_null_lon"] = np.where(df_4b["null_lon"] == 0, 1, 0)

<ipython-input-113-a49359ad36f4>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_4b["null_lon"] = np.where(df_4b["longitude"] > -200, 1, 0)
<ipython-input-113-a49359ad36f4>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_4b["not_null_lon"] = np.where(df_4b["null_lon"] == 0, 1, 0)


In [153]:
df_4b_final = df_4b[['postal_code', 'null_lon', 'not_null_lon']].groupby('postal_code').sum()
df_4b_final

null_lon  not_null_lon
postal_code                        
94102             262           196
94103             306           252
94104              74            59
94105             126           100
94107             204           247
94108             142            82
94109             237           146
94110             334           251
94111             167           115
94112             131            59
94114             122           108
94115             162            76
94116              58            35
94117             130            67
94118             143           100
94121             106            55
94122             147           123
94123             116            53
94124              76            88
94127              43            28
94131              34            14
94132              69            65
94133             282           135
94134              45            29

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [158]:
df_4b_final["proportion_missing"] = df_4b_final["null_lon"] / (df_4b_final["not_null_lon"] + df_4b_final["null_lon"])
df_4b_final = df_4b_final.sort_values(by=['proportion_missing'], ascending= False )
df_4b_final.head(5)
                                                               
                                                               

null_lon  not_null_lon  proportion_missing
postal_code                                            
94131              34            14            0.708333
94112             131            59            0.689474
94123             116            53            0.686391
94115             162            76            0.680672
94133             282           135            0.676259

In [172]:
df_4b_final = df_4b_final.reset_index()
df_4b_final["postal_code"][:5]

0    94131
1    94112
2    94123
3    94115
4    94133
Name: postal_code, dtype: object

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [177]:
def five_a():
    if ins["business_id"].nunique() == len(ins) is True:
        return len(ins)
    else:
        print("there is not one inspection record for each business")
        


In [179]:
five_a()

there is not one inspection record for each business


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [181]:
ins["type"].dtype

dtype('O')

In [196]:
ins["type"].value_counts() 

routine      15429
complaint        1
Name: type, dtype: int64

In [219]:
ins2 = ins[ins["type"].isin(["routine"])]
ins2

## would love to get help to do this in a more contained way

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [221]:
ins2 = ins2.astype({"date" : "string"})
ins2['year'] = ins2["date"].str[0:4]
ins2


business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15429 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [223]:
ins2["year"].value_counts() 

2016    5839
2014    5629
2015    3923
2013      38
Name: year, dtype: Int64

In [225]:
ins3 = ins2[ins2["year"].isin(["2016","2015", "2014"])]
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15391 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [226]:
ins4 = ins3[ins3["year"].isin(["2016"])]
ins4

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [229]:
df_6b = pd.merge(bus, ins4, how="left", on=["business_id"])
df_6b

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4              31      NORMAN'S ICE CREAM AND FREEZES   
...           ...                                 ...   
7606        90234                ACCESS SFUSD THE ARC   
7607        90236                      PHO DONG HUONG   
7608        90237                    WIN'S RESTAURANT   
7609        90238                    USA POWER MARKET   
7610        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
4              2801 LEAVENWORTH ST   San Francisco    CA       94133   
...                             ...            ...   ...         ...   
7606                1520 HOWARD ST   San Francisco    CA       94103   
7607             667 MONTEREY BLVD   San Francisco    CA       94127   
7608               3040 TARAVAL ST   San Francisco    CA       94116   
7609                1524 OCEAN AVE   San Francisco    CA       94112   
7610               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  group_by_for_2c  score      date  \
0     37.791116 -122.403816  +14154217044                1   82.0  20160503   
1     37.786848 -122.421547  +14157763262                1   94.0  20160513   
2     37.792888 -122.403135  +14156779494                1   98.0  20161005   
3     37.792888 -122.403135  +14156779494                1   96.0  20160311   
4     37.807155 -122.419004           NaN                1   84.0  20160816   
...         ...         ...           ...              ...    ...       ...   
7606        NaN         NaN           NaN                1    NaN      <NA>   
7607        NaN         NaN  +14154945048                1    NaN      <NA>   
7608        NaN         NaN  +14158606036                1    NaN      <NA>   
7609        NaN         NaN  +14153331878                1    NaN      <NA>   
7610        NaN         NaN  +14158881393                1    NaN      <NA>   

         type  year  
0     routine  2016  
1     routine  2016  
2     routine  2016  
3     routine  2016  
4     routine  2016  
...       ...   ...  
7606      NaN  <NA>  
7607      NaN  <NA>  
7608      NaN  <NA>  
7609      NaN  <NA>  
7610      NaN  <NA>  

[7611 rows x 14 columns]

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [236]:
df_6b = df_6b.sort_values(by=['score'])
df_6b_1 = df_6b[["name","address", "score"]]
print(df_6b_1.head(20))


                             name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
7324                    POKI TIME     2101 LOMBARD ST    55.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
3563                  JADE GARDEN      3608 BALBOA ST    59.0
4234    

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!